SOAP fixes

SimoneMartino98 opened the pull request:
Hi @andrewtarzia @matteobecchi !

In this PR:

- Solved #68
- Solved #62
- Solved #61
- Updated tests

#68
Unified the arguments `soap_respectpbc` and `periodic` into a single argument (`soap_respectpbc`).

#62
Removed soapkwargs.

#61
Added a new centers argument that filters the universe selection into a smaller group, containing only the atoms on which we want to compute SOAP. For this, I added a new test to ensure the stability of the selection process.

P.S.
I’m not sure if this is a case where we should use the case_data option, since we have different assertions. Of course, if you have a better solution to suggest, let me know and I’ll update it accordingly.

andrewtarzia commented on src/dynsight/_internal/soapify/saponify.py:
```diff
@@ -21,7 +21,7 @@ def saponify_trajectory(
     soap_respectpbc: bool = True,
     n_core: int = 1,
-    **soapkwargs: Any,
+    centers: str = "all",
```
What is an example of another string? How do the user define this? IS there a limit of options?

andrewtarzia replied:
Maybe add a link to the mdnalysis selection docs?

andrewtarzia commented on src/dynsight/_internal/soapify/saponify.py:
```diff
@@ -40,7 +40,11 @@ def saponify_trajectory(
             The maximum degree of spherical harmonics (option passed to the
             desired SOAP engine). Defaults to 8.
         selection : str = "all"
```
Types go in the init arguments, not in the doc string.

andrewtarzia commented on src/dynsight/_internal/soapify/saponify.py:
```diff
@@ -71,7 +75,7 @@ def saponify_trajectory(
             cutoff = 2.0
 
-            soap = saponify_trajectory(univ, cutoff)
+            soap = saponify_trajectory(univ, cutoff,soap_respectpbc=False)
```
Add space after comma.

andrewtarzia commented on src/dynsight/_internal/soapify/saponify.py:
```diff
@@ -102,8 +110,9 @@ def saponify_trajectory(
-
-    tmp_soap = soap.create(system=traj, n_jobs=n_core)
+    tmp_soap = soap.create(
+        system=traj, n_jobs=n_core, centers=[centers_list] * len(traj)
```
If the line overhangs, then each argument should be on its own line.

SimoneMartino98 replied:
ruff automatically formats this line in this way; how can i do it? Maybe by moving the `[centers_list] * len(traj)` into a variable to avoid the overhang?

andrewtarzia replied:
No no, you just add a `,` at the end and Ruff should fix it.

andrewtarzia commented on src/dynsight/_internal/soapify/saponify.py:
```diff
@@ -21,7 +21,7 @@ def saponify_trajectory(
     soap_respectpbc: bool = True,
     n_core: int = 1,
-    **soapkwargs: Any,
+    centers: str = "all",
```
This also suggests to me that you will break peoples old code?

Are we sure no other kwargs are used?

SimoneMartino98 replied:
all the kwargs come from dscribe soap function; until today, we have never used them.

andrewtarzia commented on src/dynsight/_internal/soapify/saponify.py:
```diff
@@ -148,7 +157,7 @@ def fill_soap_vector_from_dscribe(
             cutoff = 2.0
 
-            soap = saponify_trajectory(univ, cutoff)
+            soap = saponify_trajectory(univ, cutoff, soap_respectpbc=False)
```
The default behaviour was `soap_respectpbc=True`, so this is changing the results, no? Why change the recommended default?

andrewtarzia replied:
I see, because of the issue with pbc - this was not actually being taken into account?

Do we need to add a test to show this works?

SimoneMartino98 replied:
Are you talking about the test on the documentation or about a general PBC test on Dynsight?

**For the first case**
I didn't add these types of test on the documentation (i think that @matteobecchi did) but i like them, they are useful to check if the codes within the documentation are outdated or not, I'd leave them.

**For the second case**
i don't think we need a test for the PBCs; they are managed by MDAnalysis in the universe. Dynsight does not have a single line of code that calculates the PBC (except through MDA).

The only thing the user needs to worry about is loading a universe with the PBCs active. From our side, we can automatize the detection of the pbc within a universe, but i don't know, maybe it could be better to leave the choice to the user.

This is my opinion, though; if we want to structure a test for the PBC, let's do it

andrewtarzia replied:
My concern is that you have modified the codes behaviour when there are or are not PBCs, and so we want to ensure that if the user says there is, then we actually get the right outcome. Currently, we are not testing that. right?

andrewtarzia commented on src/dynsight/_internal/timesoap/timesoap.py:
```diff
@@ -43,7 +43,7 @@ def normalize_soap(
             cutoff = 2.0
 
-            soap = saponify_trajectory(univ, cutoff)
+            soap = saponify_trajectory(univ, cutoff, soap_respectpbc=False)
```
Comment above - why change the default?

SimoneMartino98 replied:
Because the `trajectory.xyz` does not contain the PBC, then it raises an error in the doctest for the codes (this is the reason why the first attempt of this PR failed). I don't know the origin of this trajectory i didn't add this check on the documentation.

andrewtarzia commented on tests/soap/test_tsoap.py:
```diff
@@ -39,8 +39,7 @@ def test_time_soap_vectors() -> None:
         soap_traj = dynsight.soap.saponify_trajectory(
-            universe=universe,
-            soaprcut=r_c,
+            universe=universe, soaprcut=r_c, soap_respectpbc=False
```
Again, why change the default?

SimoneMartino98 replied:
Because the `trajectory.xyz` does not contain the PBC, then it raises an error in the doctest for the codes (this is the reason why the first attempt of this PR failed). I don't know the origin of this trajectory i didn't add this check on the documentation.

andrewtarzia commented on src/dynsight/_internal/soapify/saponify.py:
```diff
@@ -71,7 +75,7 @@ def saponify_trajectory(
             univ = MDAnalysis.Universe(path / "trajectory.xyz")
```
Also you have not removed soapkwargs from doc string?

andrewtarzia commented on tests/soap/test_soap.py:
```diff
@@ -63,3 +64,67 @@ def test_soap_vectors() -> None:
+            centers="id 2 or id 3",
+        )
+
+        _ = dynsight.soap.fill_soap_vector_from_dscribe(
```
Maybe make a new issue with this - or maybe it is because I do not use it, but these lines of code make zero sense to me. The soap interface is weird.

SimoneMartino98 replied:
i will check them; they come from an older version of this test; at a first look they seems to be useless. I let you know...

SimoneMartino98 replied:
They are totally useless... removed

andrewtarzia commented on tests/soap/test_soap.py:
```diff
@@ -63,3 +64,67 @@ def test_soap_vectors() -> None:
+            soaplmax=8,
+            soapnmax=8,
+            soap_respectpbc=False,
+            centers="id 2 or id 3",
```
I would like to see a test that also shows that SOAP with ID 2 != ID 3 (for example, proving that this has an impact).

andrewtarzia replied:
This is not completed?

andrewtarzia replied:
Oh, i see, you did it through an assert - I would like to see it done by changing the argument `centers`.

SimoneMartino98 commented:
Maybe i can rewrite the soap test; in order to have:
- data_case that include or not the PBC.
- data_case that include different selection for the centers.

This should fix everything. Sounds good ?

andrewtarzia commented:
> Maybe i can rewrite the soap test; in order to have:
> 
> * data_case that include or not the PBC.
> * data_case that include different selection for the centers.
> 
> This should fix everything. Sounds good ?
> 
> PS -> i will rewrite them using the same approach of the rdf one; let dynsight compute SOAP the first time and then use the results as reference.

Perfect to me!

SimoneMartino98 commented:
- solved #75

SimoneMartino98 commented:
To resume:
In this PR:

- Solved #68
- Solved #62
- Solved #61
- Solved #75 
- Revisited SOAP tests (using case_data)
- Added a new toy model to test PBC variations in SOAP tests (7 balls moving within a cube)

Setting this ready for review @andrewtarzia

andrewtarzia commented on tests/analysis/rdf/conftest.py:
```diff
@@ -1,6 +1,6 @@
 import pytest
 
-from .case_data import RDFCaseData
+from tests.analysis.rdf.case_data import RDFCaseData
```
Why do this change? the relative one should work??

SimoneMartino98 replied:
For some reason the relative path in the SOAP case didn't work, so I've used the absolute one. 

To maintain the same consistency, I modified this one. 

Maybe I did something wrong during the implementation, I can try again to put the relative for SOAP and RDF

SimoneMartino98 replied:
Update: confirmed that using the relative one for SOAP raises an error

andrewtarzia replied:
I think the SOAP tests are missing __init__.py, hence it is not working. But I would try and use the relative one in this case.

andrewtarzia commented on tests/systems/balls_7_nvt.gro:
What would be cool is taking this gro, doing weird PBC crossing and rotations to produce a new gro, and confirming that you get the same SOAP spectrum for both

andrewtarzia commented on tests/systems/octahedron.xyz:
```diff
@@ -1,180 +1,180 @@
-C 9.054 8.947 10.527
-C 8.719 10.363 10.672
-C 10.105 9.414 9.558
+C1 9.379 9.812 10.129
```
Where is this needed?

SimoneMartino98 replied:
It's still used in timeSOAP test (although we should rework this too actually)

andrewtarzia replied:
Sorry typo, why is the addition of `C1` needed compared to `C`?

SimoneMartino98 replied:
I was thinking of using this system for the center selection tests (so I needed to discriminate the atoms), but then I realized that this file doesn't contain the box information, so I created a new toy model. 
The number, however, remained (...ops).

andrewtarzia reviewed (changes requested):
Minor things to update - Otherwise, great work and almost there!

SimoneMartino98 commented on tests/systems/.balls_7_nvt.xtc_offsets.lock:
yes, I forgot to git ignore that file, I will do it tomorrow

SimoneMartino98 commented:
Update:
- Reworked TimeSOAP tests (with case_data)
- Removed deprecated octahedron system 
- Removed hidden files for the new toy model 
- minor fixes (just comments)

SimoneMartino98 commented:
To avoid machine computation differences, I needed to lower the `atol` to 1e-6. However should be enough

andrewtarzia reviewed (commented):
Test the change of imports in RDF, if it does not work locally with `just check` then we move on
